File: FolderOpacity/src/main/java/com/aliucord/plugins/FolderOpacity.java
package com.aliucord.plugins;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.SeekBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.graphics.ColorUtils;
import androidx.recyclerview.widget.RecyclerView;

import com.aliucord.Logger;
import com.aliucord.entities.Plugin;
import com.aliucord.patcher.PinePatchFn;
import com.aliucord.utils.DimenUtils;
import com.aliucord.utils.ReflectUtils;
import com.aliucord.views.Divider;
import com.discord.databinding.WidgetGuildFolderSettingsBinding;
import com.discord.widgets.guilds.WidgetGuildFolderSettings;
import com.discord.widgets.guilds.WidgetGuildFolderSettings$configureUI$3;
import com.discord.widgets.guilds.WidgetGuildFolderSettingsViewModel;
import com.discord.widgets.guilds.list.FolderItemDecoration;
import com.discord.widgets.guilds.list.GuildListItem;
import com.discord.widgets.guilds.list.GuildListViewHolder;
import com.lytefast.flexinput.R;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Objects;

import d0.z.d.m;
import top.canyie.pine.Pine;
import top.canyie.pine.callback.MethodReplacement;

public class FolderOpacity extends Plugin {
    private final Logger logger = new Logger("FolderOpacity");

    @NonNull
    @Override
    public Manifest getManifest() {
        Manifest manifest = new Manifest();
        manifest.authors = new Manifest.Author[]{ new Manifest.Author("zt", 289556910426816513L) };
        manifest.description = "Adds an option to the guild folder settings to set the opacity";
        manifest.version = "1.0.0";
        manifest.updateUrl = "https://raw.githubusercontent.com/zt64/aliucord-plugins/builds/updater.json";
        return manifest;
    }

    @Override
    public void start(Context context) throws NoSuchMethodException {
        final int seekBarId = View.generateViewId();

        Method getBinding = WidgetGuildFolderSettings.class.getDeclaredMethod("getBinding");
        getBinding.setAccessible(true);

        patcher.patch(WidgetGuildFolderSettings.class.getDeclaredMethod("configureUI", WidgetGuildFolderSettingsViewModel.ViewState.class), new PinePatchFn(callFrame -> {
            try {
                WidgetGuildFolderSettings _this = (WidgetGuildFolderSettings) callFrame.thisObject;
                WidgetGuildFolderSettingsViewModel viewModel = WidgetGuildFolderSettings.access$getViewModel$p(_this);
                WidgetGuildFolderSettingsBinding binding = (WidgetGuildFolderSettingsBinding) getBinding.invoke(_this);
                Context ctx = _this.getContext();
                assert ctx != null;
                assert binding != null;

                LinearLayout linearLayout = (LinearLayout) binding.d.getParent();

                if (linearLayout.findViewById(seekBarId) != null) return;

                int opacity = settings.getInt(viewModel.getFolderId() + "opacity", 255);

                TextView opacityHeader = new TextView(ctx, null, 0, R.h.UiKit_Settings_Item_Header);
                opacityHeader.setText("Folder Opacity");

                TextView currentOpacity = new TextView(ctx, null, 0, R.h.UiKit_TextView);
                currentOpacity.setText(String.valueOf(opacity));
                currentOpacity.setWidth(DimenUtils.dpToPx(28));

                SeekBar seekbar = new SeekBar(ctx, null, 0, R.h.UiKit_SeekBar);
                seekbar.setId(seekBarId);
                seekbar.setPadding(DimenUtils.dpToPx(12), 0, DimenUtils.dpToPx(12), 0);
                seekbar.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT));
                seekbar.setMax(255);
                seekbar.setProgress(opacity);
                seekbar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
                    @Override
                    public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                        currentOpacity.setText(String.valueOf(progress));
                    }

                    @Override
                    public void onStartTrackingTouch(SeekBar seekBar) {
                        binding.f.setVisibility(View.VISIBLE);
                    }

                    @Override
                    public void onStopTrackingTouch(SeekBar seekBar) { }
                });

                LinearLayout opacitySection = new LinearLayout(ctx, null, 0, R.h.UiKit_Settings_Item);
                opacitySection.addView(currentOpacity);
                opacitySection.addView(seekbar);

                linearLayout.addView(new Divider(ctx));
                linearLayout.addView(opacityHeader);
                linearLayout.addView(opacitySection);

//                binding.f.setOnClickListener(v -> {
//                    Utils.showToast(ctx, "saved");
//                    (new WidgetGuildFolderSettings$configureUI$3(_this)).onClick(v);
//                    settings.setInt(viewModel.getFolderId() + "opacity", seekbar.getProgress());
//                });
            } catch (IllegalAccessException | InvocationTargetException e) {
                logger.error(e);
            }
        }));

        patcher.patch(WidgetGuildFolderSettings$configureUI$3.class.getDeclaredMethod("onClick", View.class), new PinePatchFn(callFrame -> {
            WidgetGuildFolderSettings widgetGuildFolderSettings = ((WidgetGuildFolderSettings$configureUI$3) callFrame.thisObject).this$0;
            WidgetGuildFolderSettingsViewModel viewModel = WidgetGuildFolderSettings.access$getViewModel$p(((WidgetGuildFolderSettings$configureUI$3) callFrame.thisObject).this$0);

            try {
                SeekBar seekBar = ((WidgetGuildFolderSettingsBinding) Objects.requireNonNull(getBinding.invoke(widgetGuildFolderSettings))).getRoot().findViewById(seekBarId);
                settings.setInt(viewModel.getFolderId() + "opacity", seekBar.getProgress());
            } catch (IllegalAccessException | InvocationTargetException e) {
                logger.error(e);
            }
        }));

        Method _drawBackgroundForInitialFolder = FolderItemDecoration.class.getDeclaredMethod("drawBackgroundForInitialFolder", Canvas.class, RecyclerView.class);
        _drawBackgroundForInitialFolder.setAccessible(true);

        patcher.patch(FolderItemDecoration.class.getDeclaredMethod("onDraw", Canvas.class, RecyclerView.class, RecyclerView.State.class), new MethodReplacement() {
            @Override
            protected Object replaceCall(Pine.CallFrame callFrame) throws Throwable {
                FolderItemDecoration _this = (FolderItemDecoration) callFrame.thisObject;

                Drawable drawableNoChildren = (Drawable) ReflectUtils.getField(_this, "drawableNoChildren");
                Drawable drawableWithChildren = (Drawable) ReflectUtils.getField(_this, "drawableWithChildren");
                Drawable tintableDrawableNoChildren = (Drawable) ReflectUtils.getField(_this, "tintableDrawableNoChildren");
                int halfSize = (int) Objects.requireNonNull(ReflectUtils.getField(_this, "halfSize"));

                Canvas canvas = (Canvas) callFrame.args[0];
                RecyclerView recyclerView = (RecyclerView) callFrame.args[1];

                Drawable drawable;
                m.checkNotNullParameter(canvas, "c");
                m.checkNotNullParameter(recyclerView, "parent");
                if (recyclerView.getChildCount() < 1) return null;

                int childCount = recyclerView.getChildCount();
                for (int drawBackgroundForInitialFolder = (int) Objects.requireNonNull(_drawBackgroundForInitialFolder.invoke(_this, canvas, recyclerView)); drawBackgroundForInitialFolder < childCount; drawBackgroundForInitialFolder++) {
                    View childAt = recyclerView.getChildAt(drawBackgroundForInitialFolder);
                    RecyclerView.ViewHolder childViewHolder = recyclerView.getChildViewHolder(childAt);
                    if (childViewHolder instanceof GuildListViewHolder.FolderViewHolder) {
                        GuildListViewHolder.FolderViewHolder folderViewHolder = (GuildListViewHolder.FolderViewHolder) childViewHolder;
                        if (!folderViewHolder.shouldDrawDecoration()) continue;

                        m.checkNotNullExpressionValue(childAt, "view");
                        int left = (childAt.getLeft() + childAt.getRight()) / 2;
                        int top = (childAt.getTop() + childAt.getBottom()) / 2;
                        int numChildren = folderViewHolder.getNumChildren();
                        if (numChildren == 0) {
                            Integer color = folderViewHolder.getColor();
                            if (color != null) {
                                assert tintableDrawableNoChildren != null;
                                int alpha = settings.getInt(((GuildListItem.FolderItem) Objects.requireNonNull(ReflectUtils.getField(folderViewHolder, "data"))).getFolderId() + "opacity", 255);
                                color = ColorUtils.setAlphaComponent(color, alpha);
                                tintableDrawableNoChildren.setColorFilter(color, PorterDuff.Mode.SRC);

                                drawable = tintableDrawableNoChildren;
                            } else {
                                drawable = drawableNoChildren;
                            }
                            assert drawable != null;
                            drawable.setBounds(left - halfSize, top - halfSize, left + halfSize, top + halfSize);
                            drawable.draw(canvas);
                        } else {
                            int height = (childAt.getHeight() * numChildren) + halfSize + top;
                            assert drawableWithChildren != null;
                            drawableWithChildren.setBounds(left - halfSize, top - halfSize, left + halfSize, height);
                            drawableWithChildren.draw(canvas);
                        }
                    }
                    if ((childViewHolder instanceof GuildListViewHolder.GuildViewHolder) && ((GuildListViewHolder.GuildViewHolder) childViewHolder).isTargetedForFolderCreation()) {
                        m.checkNotNullExpressionValue(childAt, "view");
                        int left2 = (childAt.getLeft() + childAt.getRight()) / 2;
                        int top2 = (childAt.getTop() + childAt.getBottom()) / 2;
                        assert drawableNoChildren != null;
                        drawableNoChildren.setBounds(left2 - halfSize, top2 - halfSize, left2 + halfSize, top2 + halfSize);
                        drawableNoChildren.draw(canvas);
                    }
                }
                return null;
            }
        });
    }

    @Override
    public void stop(Context context) { patcher.unpatchAll(); }
}

File: FolderOpacity/src/main/java/com/aliucord/plugins/FolderOpacity.kt
package com.aliucord.plugins

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.PorterDuffColorFilter
import android.view.View
import android.widget.*
import android.widget.SeekBar.OnSeekBarChangeListener
import androidx.core.content.ContextCompat
import androidx.core.graphics.ColorUtils
import com.aliucord.Utils
import com.aliucord.annotations.AliucordPlugin
import com.aliucord.entities.Plugin
import com.aliucord.patcher.Hook
import com.aliucord.patcher.InsteadHook
import com.aliucord.utils.DimenUtils.dp
import com.aliucord.views.Divider
import com.discord.databinding.WidgetGuildFolderSettingsBinding
import com.discord.databinding.WidgetGuildsListItemFolderBinding
import com.discord.widgets.guilds.WidgetGuildFolderSettings
import com.discord.widgets.guilds.WidgetGuildFolderSettingsViewModel
import com.discord.widgets.guilds.`WidgetGuildFolderSettings$configureUI$3`
import com.discord.widgets.guilds.list.GuildListItem
import com.discord.widgets.guilds.list.GuildListViewHolder
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.lytefast.flexinput.R
import java.util.*
import kotlin.collections.HashMap

@AliucordPlugin
class FolderOpacity : Plugin() {
    private val folderViewHolderMap = HashMap<Long, GuildListViewHolder.FolderViewHolder>()

    private val folderBindingField = GuildListViewHolder.FolderViewHolder::class.java.getDeclaredField("binding").apply { isAccessible = true }
    private val GuildListViewHolder.FolderViewHolder.binding get() = folderBindingField[this] as WidgetGuildsListItemFolderBinding

    private val getBindingMethod = WidgetGuildFolderSettings::class.java.getDeclaredMethod("getBinding").apply { isAccessible = true }
    private fun WidgetGuildFolderSettings.getBinding() = getBindingMethod.invoke(this) as WidgetGuildFolderSettingsBinding

    private val folderContainerId = Utils.getResId("guilds_item_folder_container", "id")

    private fun setAlpha(folder: GuildListViewHolder.FolderViewHolder, alpha: Int) {
        val root = folder.binding.root
        val color = ColorUtils.setAlphaComponent(folder.color ?: Color.WHITE, alpha)
        val background = ContextCompat.getDrawable(root.context, R.e.drawable_squircle_white_alpha_30)

        background?.colorFilter = PorterDuffColorFilter(color, PorterDuff.Mode.SRC)
        root.findViewById<FrameLayout>(folderContainerId).background = background
    }

    @SuppressLint("SetTextI18n")
    override fun start(context: Context) {
        val folderColorPickerId = Utils.getResId("guild_folder_settings_color", "id")
        val saveButtonId = Utils.getResId("guild_folder_settings_save", "id")
        val seekBarId = View.generateViewId()

        // Disable decoration drawing
        patcher.patch(GuildListViewHolder.FolderViewHolder::class.java.getDeclaredMethod("shouldDrawDecoration"), InsteadHook.returnConstant(false))

        // Set the background color with alpha
        patcher.patch(GuildListViewHolder.FolderViewHolder::class.java.getDeclaredMethod("configure", GuildListItem.FolderItem::class.java), Hook {
            val thisObject = it.thisObject as GuildListViewHolder.FolderViewHolder
            val folderItem = it.args[0] as GuildListItem.FolderItem

            folderViewHolderMap[folderItem.folderId] = thisObject

            setAlpha(thisObject, settings.getInt(folderItem.folderId.toString() + "opacity", 30))
        })

        // Patch folder settings
        patcher.patch(WidgetGuildFolderSettings::class.java.getDeclaredMethod("configureUI", WidgetGuildFolderSettingsViewModel.ViewState::class.java), Hook {
            val thisObject = it.thisObject as WidgetGuildFolderSettings

            val root = thisObject.getBinding().root
            val ctx = thisObject.requireContext()
            val linearLayout = root.findViewById<RelativeLayout>(folderColorPickerId).parent as LinearLayout
            if (linearLayout.findViewById<View>(seekBarId) != null) return@Hook

            val opacity = settings.getInt(WidgetGuildFolderSettings.`access$getViewModel$p`(thisObject).folderId.toString() + "opacity", 30)
            val currentOpacity = TextView(ctx, null, 0, R.i.UiKit_TextView).apply {
                text = opacity.toString()
                width = 28.dp
            }
            with(linearLayout) {
                addView(Divider(ctx))
                addView(TextView(ctx, null, 0, R.i.UiKit_Settings_Item_Header).apply {
                    text = "Folder Opacity"
                })
                addView(LinearLayout(ctx, null, 0, R.i.UiKit_Settings_Item).apply {
                    addView(currentOpacity)
                    addView(SeekBar(ctx, null, 0, R.i.UiKit_SeekBar).apply {
                        id = seekBarId
                        layoutParams = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)
                        max = 255
                        progress = opacity
                        12.dp.let { dp -> setPadding(dp, 0, dp, 0) }
                        setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
                            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                                currentOpacity.text = progress.toString()
                            }

                            override fun onStartTrackingTouch(seekBar: SeekBar) {
                                root.findViewById<FloatingActionButton>(saveButtonId).visibility = View.VISIBLE
                            }

                            override fun onStopTrackingTouch(seekBar: SeekBar) {}
                        })
                    })
                })
            }
        })

        // Patch save button
        patcher.patch(`WidgetGuildFolderSettings$configureUI$3`::class.java.getDeclaredMethod("onClick", View::class.java), Hook {
            val thisObject = it.thisObject as `WidgetGuildFolderSettings$configureUI$3`
            val widgetGuildFolderSettings = thisObject.`this$0`
            val folderId = WidgetGuildFolderSettings.`access$getViewModel$p`(widgetGuildFolderSettings).folderId
            val seekBar = widgetGuildFolderSettings.getBinding().root.findViewById<SeekBar>(seekBarId)

            settings.setInt(folderId.toString() + "opacity", seekBar.progress)

            setAlpha(folderViewHolderMap[folderId]!!, seekBar.progress)
        })
    }

    override fun stop(context: Context) = patcher.unpatchAll()
}